### ISSUE TITLE
undefined method `caching_tag_list_on?'

### ISSUE THREAD

**Issue by aaronchi** [mbleigh/acts-as-taggable-on#432]
Since #403 was applied, I'm getting an error rendering a tag_list because the caching_xxx_on? method_on method has not been added. This is rails 4.0.2 working against the acts-as-taggable-on master. It seems that using base.connected? doesn't work reliably on initialization and causes errors later if these methods have not been defined.

Stacktrace is not very interesting but here it is:

```
Showing /Users/aaron/sites/mixxter/app/views/admin/sites/_form.html.erb where line #5 raised:

undefined method `caching_tag_list_on?' for #<Class:0x007fc65e0ea6d8>
Extracted source (around line #5):

    wrapper: :bootstrap3_horizontal, defaults: {column_html: {class: "controls col-sm-9"}, label_html: {class: "col-sm-3"} },
    html: {class: "form-horizontal", role: "form"} do |f| %>
    <%= f.input :name %>
    <%= f.input :tag_list %>
    <%= f.input :description, input_html: {class: 'auto'} %>

    <div class="form-group">

Trace of template inclusion: app/views/admin/sites/edit.html.erb, app/views/admin/sites/show.html.erb

Rails.root: /Users/aaron/sites/mixxter

Application Trace | Framework Trace | Full Trace
app/views/admin/sites/_form.html.erb:5:in `block in _app_views_admin_sites__form_html_erb__4128459791633110155_70244962879300'
app/views/admin/sites/_form.html.erb:1:in `_app_views_admin_sites__form_html_erb__4128459791633110155_70244962879300'
app/views/admin/sites/edit.html.erb:9:in `_app_views_admin_sites_edit_html_erb___2528484288122667039_70244963004780'
app/views/admin/sites/show.html.erb:1:in `_app_views_admin_sites_show_html_erb___3959670009797880860_70244963057000'
app/controllers/application_controller.rb:88:in `set_timezone'
```

**Comment by kris** [comment]
This isn't fixed, unfortunately. I've come across this doing a Rails 4->5 migration and have noticed something interesting. I added some logger calls to the `ActiveRecord::Base.columns` monkey patch and found that it's being called sporadically. I'm not 100% sure why, but this is what I see in Rails console:

`Article.last.tag_list` raises an exception about the missing method
`a = Article.last; a.tag_list` will work.

I'm not sure why this is the issue, but it's what I am seeing, and can repeat 100% locally.

**Comment by courtenay** [comment]
It may be related to calling the (missing) method on the activemodel relation (foo.last.tag_list or foo.person.tag_list) as opposed to the final object (foo.person ; foo.person.tag_list )

**Comment by digerata** [comment]
I'm seeing this as well, Rails 4 -> 5 upgrade. What I've found is after the error happens, going back to the console and calling `foo.person.tag_list` directly will work and the error no longer happens for that session.

**Comment by digerata** [comment]
Gemfile version:

    acts-as-taggable-on (4.0.0)

Stack Trace:
```
Completed 500 Internal Server Error in 716ms (ActiveRecord: 91.1ms)

NoMethodError - undefined method `caching_tag_list_on?' for #<Class:0x007fa6abba7eb8>:
  activerecord (5.0.0) lib/active_record/dynamic_matchers.rb:21:in `method_missing'
  attr_encrypted (3.0.3) lib/attr_encrypted.rb:295:in `method_missing'
  attr_encrypted (3.0.3) lib/attr_encrypted/adapters/active_record.rb:129:in `method_missing_with_attr_encrypted'
  acts-as-taggable-on (4.0.0) lib/acts_as_taggable_on/taggable/core.rb:286:in `tag_list_cache_on'
  acts-as-taggable-on (4.0.0) lib/acts_as_taggable_on/taggable/core.rb:295:in `tag_list_on'
  acts-as-taggable-on (4.0.0) lib/acts_as_taggable_on/taggable/core.rb:39:in `specialization_list'
  app/views/jobs/summary/_specialization.html.erb:6:in `_app_views_jobs_summary__specialization_html_erb__715560487591950680_70176910023640'
  actionview (5.0.0) lib/action_view/template.rb:158:in `block in render'
  activesupport (5.0.0) lib/active_support/notifications.rb:166:in `instrument'
  actionview (5.0.0) lib/action_view/template.rb:348:in `instrument'
  actionview (5.0.0) lib/action_view/template.rb:156:in `render'
  actionview (5.0.0) lib/action_view/renderer/partial_renderer.rb:343:in `render_partial'
  actionview (5.0.0) lib/action_view/renderer/partial_renderer.rb:311:in `block in render'
  actionview (5.0.0) lib/action_view/renderer/abstract_renderer.rb:42:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications/instrumenter.rb:21:in `instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `instrument'
  actionview (5.0.0) lib/action_view/renderer/abstract_renderer.rb:41:in `instrument'
  actionview (5.0.0) lib/action_view/renderer/partial_renderer.rb:310:in `render'
  actionview (5.0.0) lib/action_view/renderer/renderer.rb:47:in `render_partial'
  actionview (5.0.0) lib/action_view/renderer/renderer.rb:21:in `render'
  actionview (5.0.0) lib/action_view/helpers/rendering_helper.rb:32:in `render'
  app/views/jobs/_row.html.erb:16:in `_app_views_jobs__row_html_erb__2748290585986942527_70176916820720'
  actionview (5.0.0) lib/action_view/template.rb:158:in `block in render'
  activesupport (5.0.0) lib/active_support/notifications.rb:166:in `instrument'
  actionview (5.0.0) lib/action_view/template.rb:348:in `instrument'
  actionview (5.0.0) lib/action_view/template.rb:156:in `render'
  actionview (5.0.0) lib/action_view/renderer/partial_renderer.rb:343:in `render_partial'
  actionview (5.0.0) lib/action_view/renderer/partial_renderer.rb:311:in `block in render'
  actionview (5.0.0) lib/action_view/renderer/abstract_renderer.rb:42:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications/instrumenter.rb:21:in `instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `instrument'
  actionview (5.0.0) lib/action_view/renderer/abstract_renderer.rb:41:in `instrument'
  actionview (5.0.0) lib/action_view/renderer/partial_renderer.rb:310:in `render'
  actionview (5.0.0) lib/action_view/renderer/renderer.rb:47:in `render_partial'
  actionview (5.0.0) lib/action_view/renderer/renderer.rb:21:in `render'
  actionview (5.0.0) lib/action_view/helpers/rendering_helper.rb:32:in `render'
  app/views/super/dashboard/index.html.erb:7:in `block in _app_views_super_dashboard_index_html_erb__2839787348230277349_70176916693820'
  activerecord (5.0.0) lib/active_record/relation/delegation.rb:38:in `each'
  app/views/super/dashboard/index.html.erb:6:in `_app_views_super_dashboard_index_html_erb__2839787348230277349_70176916693820'
  actionview (5.0.0) lib/action_view/template.rb:158:in `block in render'
  activesupport (5.0.0) lib/active_support/notifications.rb:166:in `instrument'
  actionview (5.0.0) lib/action_view/template.rb:348:in `instrument'
  actionview (5.0.0) lib/action_view/template.rb:156:in `render'
  actionview (5.0.0) lib/action_view/renderer/template_renderer.rb:54:in `block (2 levels) in render_template'
  actionview (5.0.0) lib/action_view/renderer/abstract_renderer.rb:42:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications/instrumenter.rb:21:in `instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `instrument'
  actionview (5.0.0) lib/action_view/renderer/abstract_renderer.rb:41:in `instrument'
  actionview (5.0.0) lib/action_view/renderer/template_renderer.rb:53:in `block in render_template'
  actionview (5.0.0) lib/action_view/renderer/template_renderer.rb:61:in `render_with_layout'
  actionview (5.0.0) lib/action_view/renderer/template_renderer.rb:52:in `render_template'
  actionview (5.0.0) lib/action_view/renderer/template_renderer.rb:14:in `render'
  actionview (5.0.0) lib/action_view/renderer/renderer.rb:42:in `render_template'
  actionview (5.0.0) lib/action_view/renderer/renderer.rb:23:in `render'
  actionview (5.0.0) lib/action_view/rendering.rb:103:in `_render_template'
  actionpack (5.0.0) lib/action_controller/metal/streaming.rb:217:in `_render_template'
  actionview (5.0.0) lib/action_view/rendering.rb:83:in `render_to_body'
  actionpack (5.0.0) lib/action_controller/metal/rendering.rb:52:in `render_to_body'
  actionpack (5.0.0) lib/action_controller/metal/renderers.rb:144:in `render_to_body'
  actionpack (5.0.0) lib/abstract_controller/rendering.rb:26:in `render'
  actionpack (5.0.0) lib/action_controller/metal/rendering.rb:36:in `render'
  actionpack (5.0.0) lib/action_controller/metal/instrumentation.rb:44:in `block (2 levels) in render'
  activesupport (5.0.0) lib/active_support/core_ext/benchmark.rb:12:in `block in ms'
  /Users/mike/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/benchmark.rb:308:in `realtime'
  activesupport (5.0.0) lib/active_support/core_ext/benchmark.rb:12:in `ms'
  actionpack (5.0.0) lib/action_controller/metal/instrumentation.rb:44:in `block in render'
  actionpack (5.0.0) lib/action_controller/metal/instrumentation.rb:87:in `cleanup_view_runtime'
  activerecord (5.0.0) lib/active_record/railties/controller_runtime.rb:25:in `cleanup_view_runtime'
  actionpack (5.0.0) lib/action_controller/metal/instrumentation.rb:43:in `render'
  remotipart (1.3.1) lib/remotipart/render_overrides.rb:16:in `render_with_remotipart'
  actionpack (5.0.0) lib/action_controller/metal/implicit_render.rb:36:in `default_render'
  actionpack (5.0.0) lib/action_controller/metal/basic_implicit_render.rb:4:in `block in send_action'
  actionpack (5.0.0) lib/action_controller/metal/basic_implicit_render.rb:4:in `send_action'
  actionpack (5.0.0) lib/abstract_controller/base.rb:188:in `process_action'
  actionpack (5.0.0) lib/action_controller/metal/rendering.rb:30:in `process_action'
  actionpack (5.0.0) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
  activesupport (5.0.0) lib/active_support/callbacks.rb:126:in `call'
  activesupport (5.0.0) lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'
  activesupport (5.0.0) lib/active_support/callbacks.rb:455:in `call'
  activesupport (5.0.0) lib/active_support/callbacks.rb:101:in `__run_callbacks__'
  activesupport (5.0.0) lib/active_support/callbacks.rb:750:in `_run_process_action_callbacks'
  activesupport (5.0.0) lib/active_support/callbacks.rb:90:in `run_callbacks'
  actionpack (5.0.0) lib/abstract_controller/callbacks.rb:19:in `process_action'
  actionpack (5.0.0) lib/action_controller/metal/rescue.rb:20:in `process_action'
  actionpack (5.0.0) lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (5.0.0) lib/active_support/notifications/instrumenter.rb:21:in `instrument'
  activesupport (5.0.0) lib/active_support/notifications.rb:164:in `instrument'
  actionpack (5.0.0) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
  actionpack (5.0.0) lib/action_controller/metal/params_wrapper.rb:248:in `process_action'
  activerecord (5.0.0) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
  actionpack (5.0.0) lib/abstract_controller/base.rb:126:in `process'
  actionview (5.0.0) lib/action_view/rendering.rb:30:in `process'
  actionpack (5.0.0) lib/action_controller/metal.rb:190:in `dispatch'
  actionpack (5.0.0) lib/action_controller/metal.rb:262:in `dispatch'
  actionpack (5.0.0) lib/action_dispatch/routing/route_set.rb:50:in `dispatch'
  actionpack (5.0.0) lib/action_dispatch/routing/route_set.rb:32:in `serve'
  actionpack (5.0.0) lib/action_dispatch/journey/router.rb:39:in `block in serve'
  actionpack (5.0.0) lib/action_dispatch/journey/router.rb:26:in `serve'
  actionpack (5.0.0) lib/action_dispatch/routing/route_set.rb:725:in `call'
  omniauth (1.3.1) lib/omniauth/strategy.rb:186:in `call!'
  omniauth (1.3.1) lib/omniauth/strategy.rb:164:in `call'
  bullet (5.4.2) lib/bullet/rack.rb:10:in `call'
  remotipart (1.3.1) lib/remotipart/middleware.rb:32:in `call'
  warden (1.2.6) lib/warden/manager.rb:35:in `block in call'
  warden (1.2.6) lib/warden/manager.rb:34:in `call'
  rack (2.0.1) lib/rack/etag.rb:25:in `call'
  rack (2.0.1) lib/rack/conditional_get.rb:25:in `call'
  rack (2.0.1) lib/rack/head.rb:12:in `call'
  rack (2.0.1) lib/rack/session/abstract/id.rb:222:in `context'
  rack (2.0.1) lib/rack/session/abstract/id.rb:216:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/cookies.rb:613:in `call'
  activerecord (5.0.0) lib/active_record/migration.rb:552:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/callbacks.rb:38:in `block in call'
  activesupport (5.0.0) lib/active_support/callbacks.rb:97:in `__run_callbacks__'
  activesupport (5.0.0) lib/active_support/callbacks.rb:750:in `_run_call_callbacks'
  activesupport (5.0.0) lib/active_support/callbacks.rb:90:in `run_callbacks'
  actionpack (5.0.0) lib/action_dispatch/middleware/callbacks.rb:36:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/executor.rb:12:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
  better_errors (2.1.1) lib/better_errors/middleware.rb:84:in `protected_app_call'
  better_errors (2.1.1) lib/better_errors/middleware.rb:79:in `better_errors_call'
  better_errors (2.1.1) lib/better_errors/middleware.rb:57:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/debug_exceptions.rb:49:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
  railties (5.0.0) lib/rails/rack/logger.rb:36:in `call_app'
  railties (5.0.0) lib/rails/rack/logger.rb:24:in `block in call'
  activesupport (5.0.0) lib/active_support/tagged_logging.rb:70:in `block in tagged'
  activesupport (5.0.0) lib/active_support/tagged_logging.rb:26:in `tagged'
  activesupport (5.0.0) lib/active_support/tagged_logging.rb:70:in `tagged'
  railties (5.0.0) lib/rails/rack/logger.rb:24:in `call'
  sprockets-rails (3.2.0) lib/sprockets/rails/quiet_assets.rb:13:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/request_id.rb:24:in `call'
  rack (2.0.1) lib/rack/method_override.rb:22:in `call'
  rack (2.0.1) lib/rack/runtime.rb:22:in `call'
  activesupport (5.0.0) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/executor.rb:12:in `call'
  actionpack (5.0.0) lib/action_dispatch/middleware/static.rb:136:in `call'
  rack (2.0.1) lib/rack/sendfile.rb:111:in `call'
  hirefire-resource (0.3.12) lib/hirefire/middleware.rb:41:in `call'
  rack-cors (0.4.0) lib/rack/cors.rb:80:in `call'
  railties (5.0.0) lib/rails/engine.rb:522:in `call'
  puma (3.6.0) lib/puma/configuration.rb:225:in `call'
  puma (3.6.0) lib/puma/server.rb:578:in `handle_request'
  puma (3.6.0) lib/puma/server.rb:415:in `process_client'
  puma (3.6.0) lib/puma/server.rb:275:in `block in run'
  puma (3.6.0) lib/puma/thread_pool.rb:116:in `block in spawn_thread'
```

**Comment by ProGM** [comment]
I made a PR to fix this (#806). Please try it and let me know if it works for you 👍 

```ruby
gem "acts_as_taggable_on", git: "https://github.com/ProGM/acts-as-taggable-on/tree", branch: 'fix-caching-tag-list-on-bug'
```

**Comment by digerata** [comment]
I can confirm your PR fixes this problem. Thanks, Piero! 👍
